src/openrct2-ui/windows/Tooltip.cpp

```diff
@@ -48,64 +136,18 @@ void WindowTooltipReset(const ScreenCoordsXY& screenCoords)
     input_set_flag(INPUT_FLAG_4, false);
 }
 
-// Returns the width of the new tooltip text
-static int32_t FormatTextForTooltip(const OpenRCT2String& message)
-{
-    utf8 tempBuffer[sizeof(gCommonStringFormatBuffer)];
-    format_string(tempBuffer, sizeof(tempBuffer), message.str, message.args.Data());
-
-    OpenRCT2String formattedMessage{ STR_STRING_TOOLTIP, Formatter() };
-    formattedMessage.args.Add<const char*>(tempBuffer);
-    format_string(_tooltipText, sizeof(_tooltipText), formattedMessage.str, formattedMessage.args.Data());
-
-    auto textWidth = gfx_get_string_width_new_lined(_tooltipText, FontStyle::Small);
-    textWidth = std::min(textWidth, 196);
-
-    int32_t numLines;
-    textWidth = gfx_wrap_string(_tooltipText, textWidth + 1, FontStyle::Small, &numLines);
-    _tooltipNumLines = numLines;
-    return textWidth;
-}
-
 void WindowTooltipShow(const OpenRCT2String& message, ScreenCoordsXY screenCoords)
 {
     auto* w = window_find_by_class(WindowClass::Error);
     if (w != nullptr)
         return;
 
-    int32_t textWidth = FormatTextForTooltip(message);
-    int32_t width = textWidth + 3;
-    int32_t height = ((_tooltipNumLines + 1) * font_get_line_height(FontStyle::Small)) + 4;
-    window_tooltip_widgets[WIDX_BACKGROUND].right = width;
-    window_tooltip_widgets[WIDX_BACKGROUND].bottom = height;
-
-    int32_t screenWidth = context_get_width();
-    int32_t screenHeight = context_get_height();
-    screenCoords.x = std::clamp(screenCoords.x - (width / 2), 0, screenWidth - width);
-
-    // TODO The cursor size will be relative to the window DPI.
-    //      The amount to offset the y should be adjusted.
-
-    int32_t max_y = screenHeight - height;
-    screenCoords.y += 26; // Normally, we'd display the tooltip 26 lower
-    if (screenCoords.y > max_y)
-        // If y is too large, the tooltip could be forced below the cursor if we'd just clamped y,
-        // so we'll subtract a bit more
-        screenCoords.y -= height + 40;
-    screenCoords.y = std::clamp(screenCoords.y, 22, max_y);
-
-    w = WindowCreate(
-        screenCoords, width, height, &window_tooltip_events, WindowClass::Tooltip, WF_TRANSPARENT | WF_STICK_TO_FRONT);
-    w->widgets = window_tooltip_widgets;
-
+    auto window = WindowCreate<TooltipWindow>(WindowClass::Tooltip, screenCoords, 0, 0, WF_TRANSPARENT | WF_STICK_TO_FRONT);
+    window->SetMessage(message, screenCoords);
```

Comment:
```suggestion
    auto tooltipWindow = TooltipWindow(message);
    auto window = WindowCreate<TooltipWindow>(std::move(tooltipWindow), WindowClass::Tooltip, screenCoords, 0, 0, WF_TRANSPARENT | WF_STICK_TO_FRONT);
```

`message` should be given during instantiation, instead of with a setter.

Reply:
Make use of constructor initialization now.
I've seen some window classes use constructor initialization and others use setter functions. Personally I don't have a preference.

Reply:
The setter function has a drawback which is to not have everything ready upon creating the window, and thus having to do the invalidates like you were doing.